devel/spicy/reference.rst

```diff
@@ -841,6 +841,102 @@ processing:
+variables into corresponding Spicy values. For example, let's say you
+would like to provide a Zeek script option with a ``count`` value that
+your Spicy analyzer can leverage. On the Zeek side, you'd define the
+option like this:
+
+.. code-block:: zeek
+
+    module MyModule;
+
+    export {
+        option my_value: count &default=42;
+    }
+
+Then, in your Spicy code, you can access the value of that option
+like this:
+
+.. code-block:: spicy
+
+    import zeek;
+
+    ...
+    local my_value = zeek::get_count("MyModule::my_value");
+    ...
+
+
+This looks up the option by its global, fully-scoped ID and returns the
+Zeek-side ``count`` as a Spicy ``uint64``. If there are any errors,
+such as the global not existing or having the wrong type, the
+``get_count()`` would abort with an exception.
+
+There are corresponding conversion functions for most of Zeek's
+built-in types, see :ref:`spicy_functions` for the full list of
+``get_<TYPE>()`` functions.
+
+For Zeek container types (``map``, ``record``, ``set``, ``vector``),
+the ``get_<TYPE>()`` functions return an opaque value that can be used
+to then inspect the container further. For example, you can check
+membership in a ``set`` like this:
+
+.. code-block:: zeek
+
+   # Zeek module `MyModule`
+   option my_set: set[count] = { 1, 2, 3 };
+
+.. code-block:: spicy
+
+   # Spicy
+   local my_set = zeek::get_set("MyModule::my_set");
+
+   if ( zeek::set_contains(my_set, 2) )
+      print "Found 2 in the set";
+
+For retrieving values from containers, there are further ``as_<TYPE>``
+functions for conversion. Example accessing a record's field:
+
+.. code-block:: zeek
+
+   # Zeek module `MyModule`
+   option my_record: record {
+       a: count &default = 42;
+       b: string &default = "foo";
+   };
+
+.. code-block:: spicy
+
+   # Spicy
+   local my_record = zeek::get_record("MyModule::my_record");
+   local a = zeek::as_count(zeek::record_field(my_record, "a"));
+   local b = zeek::as_string(zeek::record_field(my_record, "b")); # returns "bytes" (not string)
+
+See :ref:`spicy_functions` again for the full list of ``as_<TYPE>()``
+and container accessor functions.
+
+The API provides only read access to Zeek variables, there's no way to
+change them from inside Spicy. This is both to simplify the API, and
+also conceptually to avoid offering a back channel into Zeek state
+that could end up producing a very tight coupling of Spicy and Zeek
+code. Use events to communicate from Spicy to Zeek instead.
+
+Access to a Zeek global can be relatively slow as it performs the ID
+lookup and data conversion. Accordingly, the mechanism is primarily
+meant for configuration-style data, not for transferring heaps of
+dynamic state. However, as a way to speed up repeated access slightly,
+you can use :ref:`zeek::get_value <spicy_get_value>` to cache the
+result of the ID lookup, then use ``as_<TYPE>()`` on the cached value
+to retrieve the actual value.
```

Comment:
I wonder whether we should more explicitly call out a pattern on how to use Zeek values in parsers (e.g., with an dedicated example right here). A good pattern might be to capture a Zeek `option` in a Spicy `global`, and then reference that global during parsing (an example could be a value for a `&max-size` attribute). This trades repeated, potentially costly calls into Zeek with a less costly access to a `global`.

Reply:
It's a bit tricky, I'm not sure I can tell yet what a "good pattern" is. For example, caching an `option` in a global might *not* be the best way to go: options are designed to potentially change at runtime (vs consts, which don't). I think I'd prefer to leave this open for now and see how people end up using it.

Reply:
Actually, let me add couple of sentences along those lines.